# cert-revocation-checking-demo

## Scripts

### `./certificates`
- `./certificates/generate.sh` - generates keys and certificates
- `./certificates/revoke-client-cert.sh` - revokes previously generated client certificate
- `./certificates/responder.sh` - starts a local OCSP responder server at `localhost:2560`. This address is configured in certificates generated by `generate.sh`.

### `./ledger`
- `./ledger/start.sh` - creates a new DAML project with `daml new` and starts a local Sandbox instance configured to listen on `localhost:6865` over TLS. Certificate revokation checking is enabled.

### `./client-app`
- `./client-app/start.sh` - builds and starts a simple Scala client which tries to connect to the ledger using previously generated certificates.


## How to run
### Prerequisites
- A local copy of the DAML SDK with version `0.0.0` is required. Run `daml-sdk-head` which installs a version of the SDK with version number `0.0.0`.

### Steps
1. Generate certificates: `./certificates/generate.sh`
2. Start a local ledger: `./ledger/start.sh`
3. In a separate terminal start the OCSP responder: `./certificates/responder.sh`.
4. In a separate terminal start the client: `./client-app/start.sh`. The client will try to connect to the Sandbox. Sandbox will verify revocation status of client's certificate with the OCSP responder and should allow the connection.
You should get following output:
```
20:53:13.084 [default-akka.actor.default-dispatcher-8] INFO  app.ClientApp$ - Ledger API Version: 1.7.0
```
5. In a separate terminal revoke the client's certificate: `./certificates/revoke-client-cert.sh`.
6. Start the OCSP responder again. (It's deliberately configured to exit after receiving a single request to demonstrate that the Sandbox actually asks for information).
7. Start the client again. This time the connection is expected to fail due to the revoked certificate.
```
...

Exception in thread "main" io.grpc.StatusRuntimeException: UNAVAILABLE: io exception

...

Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/0:0:0:0:0:0:0:1:6865

...
```
